File: packages/eqapi_types/lib/model/v1/eew.dart
import 'package:eqapi_types/eqapi_types.dart';
import 'package:eqapi_types/lib.dart';
import 'package:eqapi_types/model/components/eew_intensity.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'eew.freezed.dart';
part 'eew.g.dart';

@freezed
class EewV1 with _$EewV1 implements V1Database {
  const factory EewV1({
    required int id,
    required int eventId,
    required String type,
    required String schemaType,
    required String status,
    required String infoType,
    required DateTime reportTime,
    int? serialno,
    String? headline,
    required bool isCanceled,
    bool? isWarning,
    required bool isLastInfo,
    DateTime? originTime,
    DateTime? arrivalTime,
    String? hypoName,
    int? depth,
    double? latitude,
    double? longitude,
    double? magnitude,
    JmaForecastIntensity? forecastMaxIntensity,
    bool? forecastMaxIntensityIsOver,
    JmaForecastLgIntensity? forecastMaxLpgmIntensity,
    bool? forecastMaxLpgmIntensityIsOver,
    List<EstimatedIntensityRegion>? regions,
    required bool? isPlum,
    required EewAccuracy? accuracy,
  }) = _EewV1;

  factory EewV1.fromJson(Map<String, dynamic> json) => _$EewV1FromJson(json);
}

@freezed
class EstimatedIntensityRegion with _$EstimatedIntensityRegion {
  const factory EstimatedIntensityRegion({
    required String code,
    required String name,
    @JsonKey(name: 'isPlum') required bool isPlum,
    @JsonKey(name: 'isWarning') required bool isWarning,
    @JsonKey(name: 'forecastMaxInt') required ForecastMaxInt forecastMaxInt,
    @JsonKey(name: 'forecastMaxLgInt')
    required ForecastMaxLgInt forecastMaxLgInt,

    /// nullの場合 `既に主要動到達と推測`
    @JsonKey(name: 'arrivalTime') required DateTime? arrivalTime,
  }) = _EstimatedIntensityRegion;

  factory EstimatedIntensityRegion.fromJson(Map<String, dynamic> json) =>
      _$EstimatedIntensityRegionFromJson(json);
}

@freezed
class EewAccuracy with _$EewAccuracy {
  @JsonSerializable(fieldRename: FieldRename.none)
  const factory EewAccuracy({
    /// ['0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8',
    /// '0' | '1' | '2' | '3' | '4' | '9']
    @JsonKey(fromJson: stringListToIntList, toJson: intListToStringList)
    required List<int> epicenters,
    @JsonKey(fromJson: int.parse, toJson: intToString) required int depth,
    @JsonKey(fromJson: int.parse, toJson: intToString)
    required int magnitudeCalcuration,
    @JsonKey(fromJson: int.parse, toJson: intToString)
    required int numberOfMagnitudeCalculation,
  }) = _EewAccuracy;

  factory EewAccuracy.fromJson(Map<String, dynamic> json) =>
      _$EewAccuracyFromJson(json);
}

String intToString(int? value) {
  if (value == null) {
    return '';
  }
  return value.toString();
}

List<int> stringListToIntList(List<String> value) =>
    value.map(int.parse).toList();
List<String> intListToStringList(List<int> value) =>
    value.map((e) => e.toString()).toList();

File: backend/packages/database/src/db/schema/eew.ts
import { sql } from "drizzle-orm";
import {
	bigint,
	boolean,
	integer,
	jsonb,
	numeric,
	pgTable,
	serial,
	smallint,
	text,
	timestamp,
} from "drizzle-orm/pg-core";
import { jmaIntensityEnum, jmaLgIntensityEnum } from "./enums";

/** 緊急地震速報 */
export const eew = pgTable("eew", {
	id: serial("id").primaryKey(),
	eventId: bigint("event_id", { mode: "bigint" }).notNull(),
	type: text("type").notNull(),
	schemaType: text("schema_type").notNull(),
	status: text("status").notNull(),
	infoType: text("info_type").notNull(),
	serialNo: integer("serial_no"),
	headline: text("headline"),
	isCanceled: boolean("is_canceled").notNull(),
	isWarning: boolean("is_warning"),
	isLastInfo: boolean("is_last_info").notNull(),
	originTime: timestamp("origin_time", { withTimezone: true }),
	arrivalTime: timestamp("arrival_time", { withTimezone: true }),
	hypoName: text("hypo_name"),
	depth: smallint("depth"),
	latitude: numeric("latitude", { precision: 3, scale: 1 }),
	longitude: numeric("longitude", { precision: 4, scale: 1 }),
	magnitude: numeric("magnitude", { precision: 2, scale: 1 }),
	forecastMaxIntensity: jmaIntensityEnum("forecast_max_intensity"),
	forecastMaxLpgmIntensity: jmaLgIntensityEnum("forecast_max_lpgm_intensity"),
	regions: jsonb("regions"),
	forecastMaxIntensityIsOver: boolean("forecast_max_intensity_is_over"),
	forecastMaxLpgmIntensityIsOver: boolean(
		"forecast_max_lpgm_intensity_is_over",
	),
	reportTime: timestamp("report_time", { withTimezone: true })
		.default(sql`now() AT TIME ZONE 'utc'`)
		.notNull(),
	accuracy: jsonb("accuracy"),
	isPlum: boolean("is_plum").notNull(),
});
